### PR TITLE
Log in with Facebook

### PR DESCRIPTION
**Pull request by joshpuetz**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adds support to logging into Forem with Facebook Oauth. We're only using the Omniauth Facebook provider, not Facebook's SDK. In addition: only the bare minimum amount of information is requested from Facebook, so no app review is required on their end (and we limit the amount of user data that we're asking for). Included is:

- a new `Authentication::Provider` subclass
- new environment variables
- migrations to add fields to `user`
- broadcast seed and production data
- of COURSE tests

## Related Tickets & Documents

resolves https://github.com/forem/InternalProjectPlanning/issues/30

## QA Instructions, Screenshots, Recordings

DEV_local__Community_👩💻👨💻

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [x] docs.forem.com
- [ ] readme
- [ ] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

The Facebook

### CODE REVIEW

File: lib/data_update_scripts/20200820153340_insert_facebook_conntect_broadcast_message.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module DataUpdateScripts
+  class InsertFacebookConntectBroadcastMessage
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply lib/data_update_scripts/20200820153340_insert_facebook_conntect_broadcast_message.rb:2]: I firmly believe we should ship with it: turning on Facebook login (or other providers in the future) is a single option in admin and the data cost of having these broadcast messages ready to go is negligible.

File: .env_sample [hunk 2]
```diff
@@ -36,6 +36,28 @@ export ELASTICSEARCH_URL="http://localhost:9200"
...
+###############################################################
+############## 3rd Party Authentication Services ##############
+###############################################################
+
+# It's best if you have the following configured
+# for more development experience
+
+# Facebook
+# (https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow)
+export FACEBOOK_APP_ID=""
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply .env_sample:48]: Oof: looks like I _kinda_ did the SiteConfig thing, but not completely. Will fix!

File: docs/backend/auth-facebook.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+---
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply docs/backend/auth-facebook.md:1]: Nope: App Secret is the important one. Regardless: this is a burner app :)

File: spec/support/fixtures/vcr_cassettes/fastly_sloan.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
+---
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply spec/support/fixtures/vcr_cassettes/fastly_sloan.yml:1]: Yup, copypasta in test setup on my part. Will fix!

**Comment by joshpuetz** [comment]
> I noticed Travis is failing due to the vendor cache missing the `omniauth-facebook` gem. If added to git I think should make that error go away and the test suite should run normally

Added!

> * If the user accepts the OAuth process without the email we should maybe halt the process (avoid creating the user) and present an error? Instead of failing (screenshot below)

This is what I'm leaning towards, although looking at the database I see that email ISN'T required. I'm curious how that works about, as both Github and Twitter require emails and appear to pass them through. Any thoughts @rhymes @benhalpern ?

**Comment by joshpuetz** [comment]
Just pushed an update that support user creation when the user opts out of sharing their email: since we still get their UID and name, I'm using that to create a unique username on our side!

File: app/services/authentication/providers/facebook.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  module Providers
+    # Facbook authentication provider, uses omniauth-facebook as backend
+    class Facebook < Provider
+      OFFICIAL_NAME = "Facebook".freeze
+      SETTINGS_URL = "https://www.facebook.com/settings?tab=applications".freeze
+
+      def new_user_data
+        {
+          name: @info.name,
+          email: @info.email || "",
```
Comment (rhymes) [inline app/services/authentication/providers/facebook.rb:11]: should we add a test for the missing email?

**Review by rhymes — COMMENTED** [review]
Nice, thanks for the updates! Left a couple of notes

**Comment by benhalpern** [comment]
I haven't examined into the weeds, but high level this all looks great. Really nice job following some of the patterns which are quite recent 👍

File: app/services/authentication/providers/facebook.rb (hunk 5 above)
Comment (joshpuetz) [reply app/services/authentication/providers/facebook.rb:11]: Yes, will do!

File: app/services/authentication/providers/facebook.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        {
+          facebook_username: @info.name,
+          facebook_created_at: Time.zone.now
+        }
+      end
+
+      # We're overriding this method because Facebook doesn't have a concept nickname or username.
+      # Instead: we'll construct one based on the user's name with some randomization thrown in based
+      # on uid, which is guaranteed to be present and unique on Facebook.
+      def user_nickname
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply app/services/authentication/providers/facebook.rb:28]: Ah, thanks: the bright red `The command "bin/knapsack_pro_rspec" exited with 1.` text was throwing me off. Imagine of Travis could color code errors properly 🤬

**Comment by leewynne** [comment]
OMG!

**Pull request merged**